tamr_toolbox/models/attribute_type.py

```diff
@@ -0,0 +1,152 @@
+"""Enum and dataclasses representing attribute types in Tamr"""
+from enum import Enum
+from dataclasses import dataclass
+from typing import ClassVar, Tuple, Union
+from tamr_toolbox.models.data_type import JsonDict
+from copy import deepcopy
+
+PrimitiveType = Enum("PrimitiveType", ["BOOLEAN", "DOUBLE", "INT", "LONG", "STRING"])
+
+# primitive type aliases
+DOUBLE = PrimitiveType.DOUBLE
+BOOLEAN = PrimitiveType.BOOLEAN
+INT = PrimitiveType.INT
+LONG = PrimitiveType.LONG
+STRING = PrimitiveType.STRING
+
+
+@dataclass(frozen=True)
+class Array:
+    """See https://docs.tamr.com/reference#attribute-types
+    NOTE:
+        `sphinx_autodoc_typehints` cannot handle forward reference to `AttributeType`,
+        so reference docs are written manually for this type
+    Args:
+        inner_type
+    """
+
+    _tag: ClassVar[str] = "ARRAY"
+    inner_type: "AttrType"
+
+
+@dataclass(frozen=True)
+class Map:
+    """See https://docs.tamr.com/reference#attribute-types
+    NOTE:
+        `sphinx_autodoc_typehints` cannot handle forward reference to `AttributeType`,
+        so reference docs are written manually for this type
+    Args:
+        inner_type
+    """
+
+    _tag: ClassVar[str] = "MAP"
+    inner_type: "AttrType"
+
+
+@dataclass(frozen=True)
+class SubAttribute:
+    """An attribute which is itself a property of another attribute.
+    See https://docs.tamr.com/reference#attribute-types
+    NOTE:
+        `sphinx_autodoc_typehints` cannot handle forward reference to `AttributeType`,
+        so reference docs are written manually for this type
+    Args:
+        name: Name of sub-attribute
+        type: See https://docs.tamr.com/reference#attribute-types
+        is_nullable: If this sub-attribute can be null
+    """
+
+    name: str
+    type: "AttrType"
+    is_nullable: bool
+
+
+@dataclass(frozen=True)
+class Record:
+    """See https://docs.tamr.com/reference#attribute-types
+    Args:
+        attributes
+    """
+
+    _tag: ClassVar[str] = "RECORD"
+    attributes: Tuple[SubAttribute, ...]
+
+
+ComplexType = Union[Array, Map, Record]
+
+AttrType = Union[PrimitiveType, ComplexType]
+
```

Comment:
I think including the DEFAULT and GEOSPATIAL aliases are going to be very useful here, so no user ever has to construct these themselves:
```suggestion
# complex type aliases
DEFAULT: AttrType = Array(STRING)
GEOSPATIAL: AttrType = Record(
    attributes=(
        SubAttribute(name="point", is_nullable=True, type=Array(DOUBLE)),
        SubAttribute(name="multiPoint", is_nullable=True, type=Array(Array(DOUBLE))),
        SubAttribute(name="lineString", is_nullable=True, type=Array(Array(DOUBLE))),
        SubAttribute(
            name="multiLineString", is_nullable=True, type=Array(Array(Array(DOUBLE)))
        ),
        SubAttribute(
            name="polygon", is_nullable=True, type=Array(Array(Array(DOUBLE)))
        ),
        SubAttribute(
            name="multiPolygon",
            is_nullable=True,
            type=Array(Array(Array(Array(DOUBLE)))),
        ),
    )
)
```